src/Proto3/Suite/JSONPB/Class.hs

```diff
@@ -0,0 +1,193 @@
+import qualified Data.Aeson.Parser                as A (eitherDecodeWith)
+import qualified Data.Aeson.Types                 as A (Object, Parser, Series,
+                                                        explicitParseFieldMaybe,
+                                                        typeMismatch)
+import qualified Data.Attoparsec.ByteString       as Atto (skipWhile)
+import qualified Data.Attoparsec.ByteString.Char8 as Atto (Parser, endOfInput)
+import qualified Data.ByteString.Lazy             as LBS
+import           Data.Proxy
+import           Data.Text                        (Text)
+import qualified Data.Text.Lazy                   as TL
+import qualified Data.Text.Lazy.Encoding          as TL
+import           Proto3.Suite.Class               (HasDefault (def, isDefault),
+                                                   Named (nameOf))
+
+-- * Typeclass definitions
+
+-- | 'A.ToJSON' variant for JSONPB direct encoding via 'A.Encoding'
+class ToJSONPB a where
+  -- | 'A.toEncoding' variant for JSONPB encoders. Equivalent to 'A.toEncoding'
+  -- if an implementation is not provided.
+  toEncodingPB :: Options -> a -> A.Encoding
+  default toEncodingPB :: (A.ToJSON a) => Options -> a -> A.Encoding
+  toEncodingPB _ = A.toEncoding
+
+-- | 'A.FromJSON' variant for JSONPB decoding from the 'A.Value' IR
+class FromJSONPB a where
+  -- | 'A.parseJSON' variant for JSONPB decoders. Equivalent to 'A.parseJSON' if
+  -- an implementation is not provided.
+  parseJSONPB :: A.Value -> A.Parser a
+
+  default parseJSONPB :: (A.FromJSON a) => A.Value -> A.Parser a
+  parseJSONPB = A.parseJSON
+
+-- * JSONPB codec entry points
+
+-- | 'Data.Aeson.encode' variant for serializing a JSONPB value as a lazy
+-- 'LBS.ByteString'.
+encode :: ToJSONPB a => Options -> a -> LBS.ByteString
+encode opts = E.encodingToLazyByteString . toEncodingPB opts
+{-# INLINE encode #-}
+
+-- | 'Data.Aeson..eitherDecode' variant for deserializing a JSONPB value from a
+-- lazy 'LBS.ByteString'.
+eitherDecode :: FromJSONPB a => LBS.ByteString -> Either String a
+eitherDecode = eitherFormatError . A.eitherDecodeWith jsonEOF (A.iparse parseJSONPB)
+  where
+    eitherFormatError = either (Left . uncurry A.formatError) Right
+    {-# INLINE eitherFormatError #-}
+
+    -- NB: cribbed from aeson-1.1.1.0:Data.Aeson.Parser.Internal.jsonEOF, which
+    -- is not exported. It's simple, so we just inline it here. Might be worth
+    -- submitting a PR to export this.
+    jsonEOF :: Atto.Parser A.Value
+    jsonEOF = A.json <* skipSpace <* Atto.endOfInput
+      where
+        skipSpace :: Atto.Parser ()
+        skipSpace = Atto.skipWhile $ \w -> w == 0x20 || w == 0x0a || w == 0x0d || w == 0x09
+        {-# INLINE skipSpace #-}
+{-# INLINE eitherDecode #-}
+
+-- * Operator definitions
+
+-- | Construct a (field name, JSONPB-encoded value) tuple
+(.=) :: (HasDefault v, ToJSONPB v) => Text -> v -> FieldPB
```

Comment:
Perhaps this could instead have type:

```haskell
(.=) :: (HasDefault v, ToJSONPB v) => Text -> v -> Option -> A.Encoding
(.=) = ... -- It uses `emit` internally to directly produce the desired encoding
```

The reason I suggest this is then you could simplify the `toEncodingPB` example to:

```haskell
toEncodingPB opts (Scalar32 i32 u32 s32 f32 sf32) = mconcat
    [ "i32"  .= i32
    , "u32"  .= u32
    , "s32"  .= s32
    , "f32"  .= f32
    , "sf32" .= sf32
    ] opts
```

... and if you changed the order of argument for `toEncodingPB` you could simplify it further to:

```haskell
toEncodingPB (Scalar32 i32 u32 s32 f32 sf32) = mconcat
    [ "i32"  .= i32
    , "u32"  .= u32
    , "s32"  .= s32
    , "f32"  .= f32
    , "sf32" .= sf32
    ]
```

Reply:
Good idea. I always forget about the `Monoid` instance for functions. 

It's slightly more involved than what you suggested, since `A.Encoding` is not a `Monoid` and we need to apply `Data.Aeson.Encoding.pairs`, but I've still simplified considerably it by swapping the argument order and getting rid of the intermediate `FieldPB` type.